### PR TITLE
GTest

### PR DESCRIPTION
**Pull request by roigcarlo**
**📝 Description**
Implementation of GTests.

### Depending on:

- [x] #12317
- [x] #12306

### Tasks

- [x] Fix exit code -11 appearing on some suites
- [x] Fix windows not executing any test 
- [x] Fix `ubuntu-intel` reporting exit code as valid, and tests reporting false positives.

### Related

- [x] #12326 (Currently Skipped)

### Kernel
Kernel changes are being discussed in separate PR's (#12281, #12306). This PR enabled the de-registration of components.

### Testing Mechanism
All cpp can now be run directly from `bin/Release/tests/[SuiteName]`. In order to tests and debug there are several new options like:
- `--gtest_filter="foo"`: To filter tests by name. Accepts wildcards.
- `--gtest_break_on_failure`: To tell to stop in the first failure
- `--gtest_catch_exceptions=0`: To prevent from capturing exepctions and be able to hook gdb directly on the faulty instruction

Complete list can be found in the gtest documentation.

Tests als can be run through the testing scripts as usual:
- `testing/run_python_tests.py`
- `testing/run_python_mpi_tests.py`
- `testing/run_cpp_tests.py`
- `testing/run_cpp_mpi_tests.py`

### Compilation Targets 
Tests are now compiled in the majority of the cases with the `kratos_add_gtests` macro. As example:

```CMAKE
## CableNet Testing sources
if(${KRATOS_BUILD_TESTING} MATCHES ON)
    file(GLOB_RECURSE KRATOS_CABLE_NET_APPLICATION_TESTING_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/tests/*.cpp)
    kratos_add_gtests(TARGET KratosCableNetCore SOURCES "${KRATOS_CABLE_NET_APPLICATION_TESTING_SOURCES}" USE_CUSTOM_MAIN)
endif(${KRATOS_BUILD_TESTING} MATCHES ON)
```

This macro accepts several options:
- `TARGET`: The target name of the library you will tests. In 99% this will be `Kratos[your_app_name]Core`. It will create a  `Kratos[your_app_name]CoreTest` as result.
- `SOURCES`: The list of the cpp files containing your tests and possible utilities needed by them. From now on we will not accept testing utilities as part of the core libraries in order to reduce the weight of the binaries.
- `USE_MPI`: Added this flag if you target depends on MPI (MPICore, Trilinos, MPIExtensions, etc...)
- `USE_CUSTOM_MAIN`: Add this flag if you have defined a custom main for your tests. In this PR this flag is enabled for most of the applications as a transition mechanism to keep the PR as small as possible. More in the transition section.

There are new several compilation targets as a result of the changes in the tests:

```
(lib) KratosCore            <- Main Kratos functionality
(lib) KratosApplication     <- Python Interface (Depends on KratosCore)
(lib) KratosTestingUtils    <- Auxiliar functions for testing (Depends on KratosCore)
(exe) KratosTests           <- The actual tests (Depends on  KratosTestingUtils, KratosCore)
(lib) KratosMPICore         <- Main MPI Kratos functionality
(lib) KratosMPITestingUtils <- Auxiliar functions for testing (Depends on KratosMPICore, KratosCore)
(exe) KratosMPITests        <- The actual tests (Depends on  KratosMPITestingUtils, KratosMPICore, KratosTestingUtils, KratosCore)
```

In general you should not bother about those unless you want to do special things with your applications.

### Transition Changes
Due to some changes in the structure of how the targets are compiler, tests no longer depend on the `testing` sources from Kratos. Specifically a change that will be required is to change "include/testing.h" to an include of you application testing suit.

As this would imply changing all testing sources from Kratos in a single PR, most of applications have a transition solution in which a generic suit is defined in the kernel and the includes are the same. A placeholder is provided as well which will need to be changed to the final solution.

StructuralMechanics, Geomechanics, CoSimulationApplication and MappingApplication have the new mechanism in which the suite is included directly from the application. We expect to replace the rest of applications in future PR once this is merged.

**🆕 Changelog**
- Adding GTests to Kratos

### CODE REVIEW

**Comment by loumalouomega** [comment]
We meet again Gtest...

**Comment by loumalouomega** [comment]
Many changes not related to Gtest in here?

File: kratos/tests/test_utilities/test_laplacian_element.cpp [hunk 1]
```diff
@@ -20,7 +20,7 @@
 #include "includes/variables.h"
 #include "utilities/math_utils.h"
 #include "utilities/geometry_utilities.h"
-#include "tests/cpp_tests/auxiliar_files_for_cpp_unnitest/test_laplacian_element.h"
+#include "tests/test_utilities/test_laplacian_element.h"
```
Comment (loumalouomega) [inline kratos/tests/test_utilities/test_laplacian_element.cpp]: Can you move these files in an independent PR?

File: scripts/wheels/linux/configure.sh [hunk 2]
```diff
@@ -75,5 +75,4 @@ cmake -H"${KRATOS_SOURCE}" -B"${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}" \
 -DINCLUDE_MMG=ON                                                       \
 -DMMG_ROOT="/workspace/external_libraries/mmg/mmg_5_5_1"               \
 -DKRATOS_BUILD_TESTING=OFF                                             \
--DINSTALL_RUNKRATOS=OFF                                                \
```
Comment (loumalouomega) [inline scripts/wheels/linux/configure.sh]: Everything related with RunKratos should go in a independent PR?

File: kratos/tests/cpp_tests/utilities/test_variable_utils.cpp [hunk 3]
```diff
@@ -59,12 +59,13 @@ KRATOS_TEST_CASE_IN_SUITE(VariableUtilsSetHistoricalVariablesToZero, KratosCoreF
...
     // Values are properly allocated
     const double tolerance = 1.0e-12;
+    const auto ZeroArray = array_1d<double, 3>(3, 0.0);
     for (const auto& r_node : r_test_model_part.Nodes()) {
-        KRATOS_CHECK_NEAR(r_node.FastGetSolutionStepValue(PRESSURE), 0.0, tolerance);
-        KRATOS_CHECK_NEAR(r_node.FastGetSolutionStepValue(TEMPERATURE), 0.0, tolerance);
-        KRATOS_CHECK_VECTOR_NEAR(r_node.FastGetSolutionStepValue(VELOCITY), ZeroVector(3), tolerance);
-        KRATOS_CHECK_VECTOR_NEAR(r_node.FastGetSolutionStepValue(DISPLACEMENT), ZeroVector(3), tolerance);
-        KRATOS_CHECK_MATRIX_NEAR(r_node.FastGetSolutionStepValue(DEFORMATION_GRADIENT), ZeroMatrix(0,0), tolerance);
+        KRATOS_EXPECT_NEAR(r_node.FastGetSolutionStepValue(PRESSURE), 0.0, tolerance);
```
Comment (loumalouomega) [inline kratos/tests/cpp_tests/utilities/test_variable_utils.cpp:64]: It was really necessary to change the checks?, it would be smarter to keep them and calling the new ones, and remove bit a bit, otherwise this comprieses many changes

File: kratos/testing_/test_skipped_exception.h [hunk 4]
```diff
@@ -0,0 +1,129 @@
+//    |  /           |
```
Comment (loumalouomega) [inline kratos/testing_/test_skipped_exception.h]: `testing_` folder?

Comment (loumalouomega) [reply kratos/testing_/test_skipped_exception.h]: I see many renamings because of this

File: kratos/tests/test_utilities/test_laplacian_element.cpp (hunk 1 above)
Comment (roigcarlo) [reply kratos/tests/test_utilities/test_laplacian_element.cpp]: I would prefer not to, but if its strictly necessary I can.

File: kratos/sources/kernel.cpp [hunk 5]
```diff
@@ -36,6 +36,36 @@ Kernel::Kernel(bool IsDistributedRun) : mpKratosCoreApplication(Kratos::make_sha
     Initialize();
 }
 
+Kernel::~Kernel() {
+    KratosComponents<Variable<bool>>::Clear();
```
Comment (loumalouomega) [inline kratos/sources/kernel.cpp]: Why this?, independent PR?

File: kratos/python/add_testing_to_python.cpp [hunk 6]
```diff
@@ -15,56 +15,56 @@
...
+// void ListOfAllTestCases() {
+//     std::cout << Testing::Tester::GetInstance() << std::endl;
+// }
 
 void  AddTestingToPython(pybind11::module& m) {
-	namespace py = pybind11;
+	// namespace py = pybind11;
 
-    py::class_<Testing::Tester, Kratos::shared_ptr<Testing::Tester> > TesterPyBind(m, "Tester");
+    // py::class_<Testing::Tester, Kratos::shared_ptr<Testing::Tester> > TesterPyBind(m, "Tester");
```
Comment (loumalouomega) [inline kratos/python/add_testing_to_python.cpp]: Remove if unused

File: kratos/tests/cpp_tests/utilities/test_variable_utils.cpp (hunk 3 above)
Comment (roigcarlo) [reply kratos/tests/cpp_tests/utilities/test_variable_utils.cpp:64]: Yep, this check specifically fails if trying to compare boost vectors (see https://github.com/KratosMultiphysics/Kratos/pull/11506/commits/8c34e8893df63d25c803ecbb542b75758b4a317b)

File: kratos/sources/kernel.cpp (hunk 5 above)
Comment (roigcarlo) [reply kratos/sources/kernel.cpp]: This is needed so the fixtures work correctly. I will detail it in the PR description

File: kratos/includes/expect.h [hunk 7]
```diff
@@ -0,0 +1,211 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
```
Comment (loumalouomega) [inline kratos/includes/expect.h:7]: Tabs!

File: kratos/tests/test_utilities/test_laplacian_element.cpp (hunk 1 above)
Comment (loumalouomega) [reply kratos/tests/test_utilities/test_laplacian_element.cpp]: This PR is humngous, it will simplify a lot review

File: kratos/tests/cpp_tests/utilities/test_variable_utils.cpp (hunk 3 above)
Comment (loumalouomega) [reply kratos/tests/cpp_tests/utilities/test_variable_utils.cpp:64]: Okay

File: kratos/sources/kernel.cpp (hunk 5 above)
Comment (loumalouomega) [reply kratos/sources/kernel.cpp]: Can be done also in a indpendent PR

Comment (loumalouomega) [reply kratos/sources/kernel.cpp]: Please otherwise will be impossible to check properly

File: kratos/tests/cpp_tests/utilities/test_variable_utils.cpp (hunk 3 above)
Comment (loumalouomega) [reply kratos/tests/cpp_tests/utilities/test_variable_utils.cpp:64]: Can we keep a transition interface?, think thatthis will break many branches and independent projects using Kratos

Comment (roigcarlo) [reply kratos/tests/cpp_tests/utilities/test_variable_utils.cpp:64]: I don't see how, I've tried. As I commented in the @KratosMultiphysics/technical-committee there is currently missuse of the `KRATOS_CHECK_XXX` macros, as they are used both as a pseudo-assertions in the tests, and as checks in the rest of the code. 

Since gtest uses full fledged assertions for the test checks, I cannot use the same macro for both, so either we change the KRATOS_CHECK in the code to something else (either be another macro or a proper check) or we make another interface for the assertions based on the existing one (what I did)

If you see any other solution tell me and I will implement it.

File: kratos/sources/kernel.cpp (hunk 5 above)
Comment (roigcarlo) [reply kratos/sources/kernel.cpp]: Agree and not. If I make a separate PR for this you will want a test, and the test can only be done if we do this. It's a snake chasing his tail :S

File: kratos/tests/test_utilities/test_laplacian_element.cpp (hunk 1 above)
Comment (roigcarlo) [reply kratos/tests/test_utilities/test_laplacian_element.cpp]: Ok I'll try, but this is only like the base of the PR. I will have to change all the apps as well. I will prepare smaller PR to merge to this one but there is no way that both systems coexists :S so changes need to be in a single PR

File: kratos/tests/cpp_tests/utilities/test_variable_utils.cpp (hunk 3 above)
Comment (loumalouomega) [reply kratos/tests/cpp_tests/utilities/test_variable_utils.cpp:64]: In that case we can change the macro from CHECK to ASSERT in a independent PR, so this PR is not 250 changes and that way other projects can be adapted

**Comment by loumalouomega** [comment]
BTW, no CTest?

**Comment by roigcarlo** [comment]
Its a draft, I already wrote in the desc that this is incomplete. So far the aim is to see of the lib is downloaded and both gtest and gmock are compiled correctly (which they seem to do). Patience.

File: kratos/testing_/test_skipped_exception.h (hunk 4 above)
Comment (roigcarlo) [reply kratos/testing_/test_skipped_exception.h]: This is leftover, I will remove it, don't worry. Its there because I still need to use things from the older testing.

File: kratos/tests/cpp_tests/utilities/test_variable_utils.cpp (hunk 3 above)
Comment (roigcarlo) [reply kratos/tests/cpp_tests/utilities/test_variable_utils.cpp:64]: The problem is that it will still be a 250+ files changed PR, but as you like.

Comment (loumalouomega) [reply kratos/tests/cpp_tests/utilities/test_variable_utils.cpp:64]: Yes, but it will be a simple change-> Just renaming macro. Here we have renaming + gtest + runkratos + ...

**Comment by roigcarlo** [comment]
@markelov208 @avdg81, I am merging the master and some of the most recent changes directly to this branch in preparation for the next week. 

If you don't mind, I will also move the changes from #11896 (specially the ones done be @markelov208  In the geo) to this branch as I have a better grip on the changes so we can have a functional basis to solve the mpi issues.

**Comment by roigcarlo** [comment]
@rfaasse @avdg81 Could you check that the last commit did not break your CI? I have changed the way tests are compiled to applications by default inherit the main that initialized the kernel instead of using the default `gmok` one. I remember you having problems with this with 0 tests detected.

Should not affect but just in case.

**Comment by rfaasse** [comment]
> @rfaasse @avdg81 Could you check that the last commit did not break your CI? I have changed the way tests are compiled to applications by default inherit the main that initialized the kernel instead of using the default `gmok` one. I remember you having problems with this with 0 tests detected.
> 
> Should not affect but just in case.

@roigcarlo Thanks for making these changes, on our pipeline the build and tests pass. For our local build, we need to deal with the dlls being in the bin folder instead of libs, but that shouldn't block this PR. 

Since @avdg81 and I are back again this week, we would like to have a meeting tomorrow afternoon to discuss which steps still need to be done to get this through!

**Comment by roigcarlo** [comment]
Perfect!, Meeting is ok as long as its before 16:30, feel free to send me the invite :+1:

**Comment by roigcarlo** [comment]
@rfaasse @avdg81 Could you check why the windows script is not running any tests :S it works on the machines I've tested (probably a problem with the path separator)

File: applications/CoSimulationApplication/tests/cpp_distributed_tests/co_simulation_distributed_suite.cpp [hunk 8]
```diff
@@ -0,0 +1,27 @@
+// KRATOS  / ___|___/ ___|(_)_ __ ___  _   _| | __ _| |_(_) ___  _ ___
```
Comment (RiccardoRossi) [inline applications/CoSimulationApplication/tests/cpp_distributed_tests/co_simulation_distributed_suite.cpp:1]: i understand that  these modifications have been agreed with  @philbucher correct?

File: applications/ConstitutiveLawsApplication/tests/cpp_tests/constitutive_laws_fast_suite.h [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+#include "testing/testing.h"
+
+namespace Kratos::Testing {
+
+class KratosConstitutiveLawsFastSuite : public KratosCoreFastSuite {
+public:
+  KratosConstitutiveLawsFastSuite();
+
+private:
+  KratosConstitutiveLawsApplication::Pointer mpConstitutiveLawsApp;
```
Comment (RiccardoRossi) [inline applications/ConstitutiveLawsApplication/tests/cpp_tests/constitutive_laws_fast_suite.h:25]: why do we need this but not the StructuralMechanics?

File: applications/ContactStructuralMechanicsApplication/tests/cpp_tests/contact_structural_mechanics_fast_suite.cpp [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+#include "contact_structural_mechanics_application.h"
+
+int main(int argc, char* argv[]) 
+{
+    ::testing::InitGoogleTest(&argc, argv);
+
+    Kratos::Testing::mApplicationInitializerList.push_back([](std::vector<Kratos::KratosApplication::Pointer> & rRegisteredApplications, Kratos::Kernel & rKernel) {
+      if (!rKernel.IsImported("ContactStructuralMechanicsApplication")) {
+        auto pApplication = std::make_shared<Kratos::KratosContactStructuralMechanicsApplication>();
+        rKernel.ImportApplication(pApplication);
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/tests/cpp_tests/contact_structural_mechanics_fast_suite.cpp:28]: i am surprised that the StructuralMEchanics is not required by these tests.

File: applications/IgaApplication/tests/cpp_tests/test_iga_shell_3p.cpp [hunk 11]
```diff
@@ -50,6 +50,8 @@ namespace Testing
     // Tests the stiffness matrix of the Shell3pElement with a polynomial degree of p=3.
     KRATOS_TEST_CASE_IN_SUITE(IgaShell3pElementP3, KratosIgaFastSuite)
     {
+        KRATOS_SKIP_TEST_IF_NOT(KratosComponents<ConstitutiveLaw>::Has("LinearElasticPlaneStress2DLaw")) << "This test needs the StructuralMechanicsApplication" << std::endl;
```
Comment (RiccardoRossi) [inline applications/IgaApplication/tests/cpp_tests/test_iga_shell_3p.cpp:53]: as i understand this means that these tests are currently being skipped. @rickyaristio @NickNick9 please take into account that they need to be moved to python ASAP

File: kratos/tests/cpp_tests/testing/test_testing.cpp.disabled
Comment (RiccardoRossi) [inline kratos/tests/cpp_tests/testing/test_testing.cpp.disabled:1]: this is intentional or is it a leftover?

**Review by RiccardoRossi — COMMENTED** [review]
some comments. I cannot claim this is a detailed review, nevertheless this looks good to me

File: kratos/tests/cpp_tests/testing/test_testing.cpp.disabled
Comment (roigcarlo) [reply kratos/tests/cpp_tests/testing/test_testing.cpp.disabled:1]: Leftover. Will fix

File: applications/IgaApplication/tests/cpp_tests/test_iga_shell_3p.cpp (hunk 11 above)
Comment (roigcarlo) [reply applications/IgaApplication/tests/cpp_tests/test_iga_shell_3p.cpp:53]: this is being tracked in https://github.com/KratosMultiphysics/Kratos/issues/12326

**Comment by jcotela** [comment]
@roigcarlo I have pushed the changes for the issue we checked earlier. Since I had added the test to the wrong suite originally, I just moved the test to the standard core suite and removed the "extra" suite class.

**Comment by roigcarlo** [comment]
Hi @KratosMultiphysics/technical-committee can someone make a final review to this? As far as I am concern seems to be working well for all the CI external to CIMNE as well as our in house CI.

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee approves this PR as it passes all the tests and provides mechanisms to handle diverse usage of the cpp tests in Kratos.

**Review by rubenzorrilla — APPROVED** [review]

**Comment by loumalouomega** [comment]
Gtest won...
